File: cpp/test/Ice/timeout/TestI.cpp
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

#include <TestI.h>
#include "Ice/Ice.h"

#include <thread>
#include <chrono>

using namespace std;
using namespace Ice;

void
TimeoutI::op(const Ice::Current&)
{
}

void
TimeoutI::sendData(Test::ByteSeq, const Ice::Current&)
{
}

void
TimeoutI::sleep(int32_t to, const Ice::Current&)
{
    this_thread::sleep_for(chrono::milliseconds(to));
}

ControllerI::ControllerI(const Ice::ObjectAdapterPtr& adapter) : _adapter(adapter) {}

void
ControllerI::holdAdapter(int32_t to, const Ice::Current&)
{
    _adapter->hold();

    if (to >= 0)
    {
        std::thread activateThread(
            [this, to]
            {
                this_thread::sleep_for(chrono::milliseconds(to));
                _adapter->activate();
            });
        activateThread.detach();
    }
}

void
ControllerI::resumeAdapter(const Ice::Current&)
{
    _adapter->activate();
}

void
ControllerI::shutdown(const Ice::Current&)
{
    _adapter->getCommunicator()->shutdown();
}

File: swift/test/Ice/timeout/TestI.swift
//
import Foundation
// Copyright (c) ZeroC, Inc. All rights reserved.
//
import Ice

class TimeoutI: Timeout {
    func op(current _: Current) throws {}

    func sendData(seq _: ByteSeq, current _: Current) throws {}

    func sleep(to: Int32, current _: Current) throws {
        Thread.sleep(forTimeInterval: TimeInterval(to) / 1000)
    }
}

class ControllerI: Controller {
    var _adapter: Ice.ObjectAdapter

    init(_ adapter: Ice.ObjectAdapter) {
        _adapter = adapter
    }

    func holdAdapter(to: Int32, current: Ice.Current) throws {
        _adapter.hold()
        if to >= 0 {
            Task {
                do {
                    self._adapter.waitForHold()
                    try await Task.sleep(for: .milliseconds(Int(to)))
                    try self._adapter.activate()
                } catch {
                    fatalError("unexpected error: \(error)")
                }

            }
        }
    }

    func resumeAdapter(current _: Ice.Current) throws {
        try _adapter.activate()
    }

    func shutdown(current: Ice.Current) {
        current.adapter.getCommunicator().shutdown()
    }
}
